File: src/de/hsh/dennis/model/NpcLogic/actors/Npc.java
package de.hsh.dennis.model.NpcLogic.actors;

import common.actor.Actor;
import common.config.WindowConfig;
import common.util.Logger;
import de.hsh.dennis.model.NpcLogic.NPCEnums;
import de.hsh.dennis.model.NpcLogic.SkinConfig;
import javafx.scene.image.Image;

public abstract class Npc extends Actor implements Comparable {

    private final double           spawnTime;
    private       double           speed = SkinConfig.Level.speed;
    private final NPCEnums.Spawn   spawnType;
    private final NPCEnums.NpcType npcType;

    Npc( final String picturePath, NPCEnums.Spawn spawnType, NPCEnums.NpcType type ) {
        super( picturePath );
        this.spawnType = spawnType;
        this.npcType = type;
        spawnTime = 0;

        switch ( type ) {
            case PACKAGE:
                setPosX( (spawnType == NPCEnums.Spawn.RIGHT) ? WindowConfig.window_width :
                         0 - SkinConfig.Package.skin_standard.getWidth() );
                setPosY( SkinConfig.Player.posY );
                break;
            case BOT:
                setPosX( (spawnType == NPCEnums.Spawn.RIGHT) ? WindowConfig.window_width :
                         0 - SkinConfig.Bot.skin_standard.getWidth() );
                setPosY( SkinConfig.Player.posY + (SkinConfig.Player.skin_standard.getHeight() * 0.2) );
                break;
            case HACKER:
                setPosX( (spawnType == NPCEnums.Spawn.RIGHT) ? WindowConfig.window_width :
                         0 - SkinConfig.Hacker.skin_standard.getWidth() );
                setPosY( SkinConfig.Player.posY + SkinConfig.Player.skin_standard.getHeight() -
                         SkinConfig.Hacker.skin_standard.getHeight() );
                break;

            default:
                Logger.log( this.getClass() + "Switching Default!" );
        }

    }

    Npc( final String picturePath, NPCEnums.Spawn spawnType, NPCEnums.NpcType type, double spawnTime, double speed ) {
        super( picturePath );
        this.spawnTime = spawnTime;
        this.npcType = type;
        this.spawnType = spawnType;
        setSpeed( speed );

        switch ( type ) {
            case PACKAGE:
                setPosX( (spawnType == NPCEnums.Spawn.RIGHT) ? WindowConfig.window_width :
                         0 - SkinConfig.Package.skin_standard.getWidth() );
                setPosY( SkinConfig.Player.posY );
                break;
            case BOT:
                setPosX( (spawnType == NPCEnums.Spawn.RIGHT) ? WindowConfig.window_width :
                         0 - SkinConfig.Bot.skin_standard.getWidth() );
                setPosY( SkinConfig.Player.posY + (SkinConfig.Player.skin_standard.getHeight() * 0.2) );
                break;
            case HACKER:
                setPosX( (spawnType == NPCEnums.Spawn.RIGHT) ? WindowConfig.window_width :
                         0 - SkinConfig.Hacker.skin_standard.getWidth() );
                setPosY( SkinConfig.Player.posY + SkinConfig.Player.skin_standard.getHeight() -
                         SkinConfig.Hacker.skin_standard.getHeight() );
                break;

            default:
                Logger.log( this.getClass() + "Switching Default!" );
        }

    }

    public void move() {
        setPosX( (spawnType == NPCEnums.Spawn.RIGHT) ? (getPosX() - getSpeed()) : (getPosX() + getSpeed()) );

    }

    @Override
    public String toString() {
        return "\nClassName:\t" + this.getClass().getName() + "\n" +
               "spawnTime:\t" + spawnTime + "\n" +
               "speed:\t\t" + speed + "\n" +
               "spawnType:\t" + spawnType.toString() + "\n" +
               "npcType:\t" + npcType.toString() + "\n"
                ;
    }

    @Override
    public int compareTo( Object o ) {
        Npc other = (Npc) o;

        if ( this.getSpawnTime() < other.getSpawnTime() ) {
            return -1;
        }
        else if ( this.getSpawnTime() > other.getSpawnTime() ) {
            return 1;
        }
        return 0;
    }

    // --- Getter & Setter ------------------------------------------------------------------------

    public Image getImage() {
        return this.getCurrentImage();
    }

    public double getPosX() {
        return this.getX();
    }

    void setPosX( double positionX ) {
        this.setX( positionX );
    }

    public double getPosY() {
        return this.getY();
    }

    void setPosY( double positionY ) {
        this.setY( positionY );
    }

    public NPCEnums.Spawn getSpawnType() {
        return spawnType;
    }

    public NPCEnums.NpcType getNpcType() {
        return npcType;
    }

    public double getSpawnTime() {
        return spawnTime;
    }
}

File: Games/Dennis/src/de/hsh/dennis/model/NpcLogic/actors/Npc.kt
package de.hsh.dennis.model.NpcLogic.actors

import common.actor.Actor
import common.config.WindowConfig
import de.hsh.dennis.model.NpcLogic.NPCEnums.NpcType
import de.hsh.dennis.model.NpcLogic.NPCEnums.Spawn
import de.hsh.dennis.model.NpcLogic.SkinConfig

abstract class Npc : Actor, Comparable<Npc> {
    val spawnTime: Double
    val spawnType: Spawn
    val npcType: NpcType

    internal constructor(picturePath: String?, spawnType: Spawn, type: NpcType) : super(picturePath!!) {
        this.spawnType = spawnType
        npcType = type
        spawnTime = 0.0
        this.speed = SkinConfig.Level.speed
        when (type) {
            NpcType.PACKAGE -> {
                x = if (spawnType == Spawn.RIGHT) WindowConfig.window_width.toDouble() else - SkinConfig.Package.skin_width.toDouble()
                y = SkinConfig.Player.posY
            }
            NpcType.BOT     -> {
                x = if (spawnType == Spawn.RIGHT) WindowConfig.window_width.toDouble() else - SkinConfig.Bot.skin_width.toDouble()
                y = SkinConfig.Player.posY + SkinConfig.Player.skin_height * 0.2
            }
            NpcType.HACKER  -> {
                x = if (spawnType == Spawn.RIGHT) WindowConfig.window_width.toDouble() else - SkinConfig.Hacker.skin_width.toDouble()
                y = SkinConfig.Player.posY + SkinConfig.Player.skin_height -
                    SkinConfig.Hacker.skin_height
            }
        }
    }

    internal constructor(picturePath: String?,
                         spawnType: Spawn,
                         type: NpcType,
                         spawnTime: Double,
                         speed: Double) : super(picturePath!!) {
        this.spawnTime = spawnTime
        npcType = type
        this.spawnType = spawnType
        this.speed = speed
        when (type) {
            NpcType.PACKAGE -> {
                x = if (spawnType == Spawn.RIGHT) WindowConfig.window_width.toDouble() else -SkinConfig.Package.skin_width.toDouble()
                y = SkinConfig.Player.posY
            }
            NpcType.BOT     -> {
                x = if (spawnType == Spawn.RIGHT) WindowConfig.window_width.toDouble() else -SkinConfig.Bot.skin_width.toDouble()
                y = SkinConfig.Player.posY + SkinConfig.Player.skin_height * 0.2
            }
            NpcType.HACKER  -> {
                x = if (spawnType == Spawn.RIGHT) WindowConfig.window_width.toDouble() else -SkinConfig.Hacker.skin_width.toDouble()
                y = SkinConfig.Player.posY + SkinConfig.Player.skin_height -
                    SkinConfig.Hacker.skin_height
            }
        }
    }

    open fun move() {
        x = if (spawnType == Spawn.RIGHT) x - speed else x + speed
    }

    override fun toString(): String {
        return """
            
            ClassName:	${this.javaClass.name}
            spawnTime:	$spawnTime
            speed:		$speed
            spawnType:	$spawnType
            npcType:	$npcType
            
            """.trimIndent()
    }

    override fun compareTo(other: Npc): Int = when {
        spawnTime < other.spawnTime -> -1
        spawnTime > other.spawnTime -> 1
        else                        -> 0
    }

}
